subsys/bluetooth/host/classic/l2cap_br.c

```diff
@@ -299,1706 +417,5209 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+				 * recovery. The received I-frame shall be discarded.
+				 */
+				discard = true;
+				LOG_DBG("Set REJ Exception");
+				if (!atomic_test_and_set_bit(br_chan->flags,
+							     L2CAP_FLAG_SEND_FRAME_REJ)) {
+					LOG_DBG("Set REJ Exception received flag");
+					atomic_set_bit(br_chan->flags,
+						       L2CAP_FLAG_SEND_FRAME_REJ_CHANGED);
+					err = l2cap_br_send_s_frame(br_chan, BT_L2CAP_CONTROL_S_RR,
+								    K_NO_WAIT);
+					if (err) {
+						LOG_ERR("Fail to send frame %d on %p", err,
+							br_chan);
+						bt_l2cap_chan_disconnect(&br_chan->chan);
+					}
+				}
+			}
+		} else {
+			/* An invalid TxSeq value is a value that does not meet either of
+			 * the above conditions and TxSeq is not equal to ExpectedTxSeq.
+			 * An I-frame with an invalid TxSeq is likely to have errors in
+			 * the control field and shall be silently discarded.
+			 */
+			return;
 		}
+	}
 
+	err = bt_l2cap_br_update_req_seq(br_chan, req_seq, false);
+	if (err) {
 		return;
 	}
 
-	if (!chan->conn->encrypt) {
+	if (discard) {
 		return;
 	}
 
-	/*
-	 * For incoming connection state send confirming outstanding
-	 * response and initiate configuration request.
-	 */
-	if (l2cap_br_conn_req_reply(chan, BT_L2CAP_BR_SUCCESS) == 0) {
-		bt_l2cap_br_chan_set_state(chan, BT_L2CAP_CONFIG);
-		/*
-		 * Initialize config request since remote needs to know
-		 * local MTU segmentation.
-		 */
-		l2cap_br_conf(chan);
-	} else if (atomic_test_and_clear_bit(BR_CHAN(chan)->flags,
-					     L2CAP_FLAG_CONN_PENDING)) {
-		buf = bt_l2cap_create_pdu(&br_sig_pool, 0);
+	if (br_chan->rx.mode == BT_L2CAP_BR_LINK_MODE_ERET) {
+		if (f && !atomic_test_and_clear_bit(br_chan->flags, L2CAP_FLAG_REJ_ACTIONED)) {
+			/* Retransmit I-frames and Send-Pending-I-frames */
+			atomic_set_bit(br_chan->flags, L2CAP_FLAG_NEW_I_FRAME);
+			atomic_set_bit(br_chan->flags, L2CAP_FLAG_RET_I_FRAME);
+			atomic_set_bit(br_chan->flags, L2CAP_FLAG_REQ_SEQ_UPDATED);
+		}
+	}
 
-		hdr = net_buf_add(buf, sizeof(*hdr));
-		hdr->code = BT_L2CAP_CONN_REQ;
-		hdr->ident = l2cap_br_get_ident();
-		hdr->len = sys_cpu_to_le16(sizeof(*req));
+valid_frame:
+	switch (sar) {
+	case BT_L2CAP_CONTROL_SAR_UNSEG:
+		__fallthrough;
+	case BT_L2CAP_CONTROL_SAR_START:
+		if (buf->len < 2) {
+			LOG_WRN("Invalid SDU length");
+			bt_l2cap_chan_disconnect(&br_chan->chan);
+			return;
+		}
+		break;
+	case BT_L2CAP_CONTROL_SAR_END:
+		__fallthrough;
+	case BT_L2CAP_CONTROL_SAR_CONTI:
+		break;
+	}
 
-		req = net_buf_add(buf, sizeof(*req));
-		req->psm = sys_cpu_to_le16(BR_CHAN(chan)->psm);
-		req->scid = sys_cpu_to_le16(BR_CHAN(chan)->rx.cid);
+	/* Redirect to experimental API. */
+	IF_ENABLED(CONFIG_BT_L2CAP_SEG_RECV, ({
+		if (br_chan->chan.ops->seg_recv) {
+			bt_l2cap_br_recv_seg_direct(br_chan, buf, sar);
+			goto done;
+		}
+	}))
+
+	if (br_chan->chan.ops->alloc_buf) {
+		err = bt_l2cap_br_recv_seg(br_chan, buf, sar);
+		if (err == -ENOBUFS) {
```

Comment:
Just a small note about code flow: you seem to have a tendency to structure the functions so that you end up with some quite deep nesting. In many places you can avoid that by inverting conditions and doing early returns or gotos. E.g. in this case this could be:
```c
	if (err != -ENOBUFS) {
		goto done;
	}
```

There are many more places in your PR like this, but it's a huge PR so I won't go pointing them out individually.

Reply:
The change can be useful for reducing one indentation for each line of code block. But It will introduce a additional `goto done;`. This change will cause doubts. From the perspective of reading, anti-logic is not conducive to reading.

Reply:
> There are many more places in your PR like this, but it's a huge PR so I won't go pointing them out individually.

Sure I will check all changes and reduce the branch nesting.

Reply:
> anti-logic is not conducive to reading

I realize this is a pretty subjective topic, but it's a style that I and likely many others coming from the Linux kernel world are used to.

Reply:
Well. It has been updated. And I have checked and fixed the similar issue.